asterius/src/Asterius/GHCi/Internals.hs

```diff
@@ -0,0 +1,415 @@
+asteriusWriteIServ hsc_env i a
+  | Just HRefl <- eqTypeRep (typeOf a) (typeRep @GHC.Msg) = case a of
+    GHC.Msg msg@(GHC.RunTH st q ty loc) -> do
+      GHC.debugTraceMsg (GHC.hsc_dflags hsc_env) 3 $ GHC.text $ show msg
+      modifyMVar_ globalGHCiState $ \s -> do
+        let run_q_exp_sym =
+              ghciClosureSymbol hsc_env "Asterius.GHCi" "asteriusRunQExp"
+            run_q_pat_sym =
+              ghciClosureSymbol hsc_env "Asterius.GHCi" "asteriusRunQPat"
+            run_q_type_sym =
+              ghciClosureSymbol hsc_env "Asterius.GHCi" "asteriusRunQType"
+            run_q_dec_sym =
+              ghciClosureSymbol hsc_env "Asterius.GHCi" "asteriusRunQDec"
+            run_q_annwrapper_sym =
+              ghciClosureSymbol hsc_env "Asterius.GHCi" "asteriusRunAnnWrapper"
+            run_mod_fin_sym =
+              ghciClosureSymbol hsc_env "Asterius.GHCi" "asteriusRunModFinalizers"
+            this_id = remoteRefToInt q
+            (sym, m) = ghciCompiledCoreExprs s IM.! this_id
+        (_, final_m, link_report) <- linkExeInMemory LinkTask
+          { progName = "",
+            linkOutput = "",
+            linkObjs = [],
+            linkLibs = [],
+            linkModule = m <> M.foldr' (<>) (ghciLibs s) (ghciObjs s),
+            debug = False,
+            gcSections = True,
+            binaryen = False,
+            verboseErr = True,
+            outputIR = Nothing,
+            rootSymbols =
+              [ run_q_exp_sym,
+                run_q_pat_sym,
+                run_q_type_sym,
+                run_q_dec_sym,
+                run_q_annwrapper_sym,
+                run_mod_fin_sym,
+                sym
+              ],
+            exportFunctions = []
+          }
+        js_s <-
+          maybe (fail "asteriusWriteIServ: RunTH: no JSSession") pure $
+            (\(x, _, _) -> x)
+              <$> ghciJSSession s
+        v <-
+          asteriusRunTH hsc_env
+            i
+            st
+            (fromIntegral (staticsSymbolMap link_report ! sym))
+            ty
+            loc
+            js_s
+            (final_m, link_report)
+        pure
+          s
+            { ghciCompiledCoreExprs = IM.delete this_id $ ghciCompiledCoreExprs s,
+              ghciJSSession =
+                let Just (x, y, _) = ghciJSSession s
+                 in Just (x, y, v)
+            }
+    GHC.Msg m@GHC.RunModFinalizers {} -> do
+      GHC.debugTraceMsg (GHC.hsc_dflags hsc_env) 3 $ GHC.text $ show m
+      withMVar globalGHCiState $ \s -> do
+        (js_s, v) <-
+          maybe (fail "asteriusWriteIServ: RunTH: no JSSession") pure $
+            (\(x, _, v) -> (x, v))
+              <$> ghciJSSession s
+        asteriusRunModFinalizers hsc_env js_s v
+    GHC.Msg m -> fail $ "asteriusWriteIServ: unsupported message " <> show m
+  | otherwise = withMVar globalGHCiState $
+    \s -> let Just (_, p, _) = ghciJSSession s in writePipe p $ put a
+
+asteriusRunTH ::
```

Comment:
Maybe use a datatype to pass all the arguments (and give them a name by doing so).

Reply:
This makes sense. Will do shortly.